lisa/transformers/kernel_installer.py

```diff
@@ -0,0 +1,308 @@
+        runbook: Any,
+        node: RemoteNode,
+        parent_log: Logger,
+        *args: Any,
+        **kwargs: Any,
+    ) -> None:
+        super().__init__(runbook, node, parent_log, *args, **kwargs)
+        self.repo_url = "http://archive.ubuntu.com/ubuntu/"
+
+    @classmethod
+    def type_name(cls) -> str:
+        return "repo"
+
+    @classmethod
+    def type_schema(cls) -> Type[schema.TypedSchema]:
+        return RepoInstallerSchema
+
+    def validate(self) -> None:
+        assert isinstance(self._node.os, Ubuntu), (
+            f"The '{self.type_name()}' installer only support Ubuntu. "
+            f"The current os is {self._node.os.__class__.__name__}"
+        )
+
+    def install(self) -> None:
+        runbook: RepoInstallerSchema = self.runbook
+        node: RemoteNode = self._node
+        ubuntu: Ubuntu = cast(Ubuntu, node.os)
+        release = node.os.os_version.codename
+
+        assert (
+            release
+        ), f"cannot find codename from the os version: {node.os.os_version}"
+
+        # add the repo
+        if runbook.is_proposed:
+            version_name = f"{release}-proposed"
+        else:
+            version_name = release
+        repo_entry = (
+            f"deb {self.repo_url} {version_name} "
+            f"restricted main multiverse universe"
+        )
+        ubuntu.wait_running_package_process()
+        result = node.execute(f'add-apt-repository "{repo_entry}"', sudo=True)
+        if result.exit_code != 0:
+            result.assert_exit_code(
+                message="failed on add repo\n"
+                + "\n".join(ubuntu.get_apt_error(result.stdout))
+            )
+
+        full_package_name = f"{runbook.source}/{version_name}"
+        self._log.info(f"installing kernel package: {full_package_name}")
+        ubuntu.install_packages(full_package_name)
+
+        kernel_version = self._get_kernel_version(runbook.source, node)
+
+        self._replace_boot_entry(kernel_version, node)
+
+        # install tool packages
+        ubuntu.install_packages(
+            [
+                f"linux-tools-{kernel_version}-azure",
+                f"linux-cloud-tools-{kernel_version}-azure",
+                f"linux-headers-{kernel_version}-azure",
+            ]
+        )
+
+    def _get_kernel_version(self, source: str, node: RemoteNode) -> str:
+        # get kernel version from apt packages
+        # linux-azure-edge/focal-proposed,now 5.11.0.1011.11~20.04.10 amd64 [installed]
+        # output: 5.11.0.1011
+        # linux-image-4.18.0-1025-azure/bionic-updates,bionic-security,now
+        # 4.18.0-1025.27~18.04.1 amd64 [installed]
+        # output 4.18.0-1025
+        kernel_version_package_pattern = re.compile(
+            f"^{source}/[^ ]+ "
+            r"(?P<kernel_version>[^.]+\.[^.]+\.[^.-]+[.-][^.]+)\..*[\r\n]+",
+            re.M,
+        )
+        # the grep is used to remove color code
+        result = node.execute(f"apt search {source} | grep ''", shell=True)
```

Comment:
I thought we were filtering out all ansi escape chars for people for free already, why bother?

Reply:
It applies to log only. I will move it to a method, and can support more. But it shouldn't apply to all stdout, since the raw content may be needed in some cases.

Reply:
Oh, true. Now I remember I apply to stdout by hand in my case.